hexrdgui/mask_manager.py

```diff
@@ -0,0 +1,281 @@
+    def max_val(self):
+        return self._max
+
+    @max_val.setter
+    def max_val(self, val):
+        self._max = val
+
+    def get_data(self):
+        return [self.min_val, self.max_val]
+
+    def set_data(self, data):
+        self.min_val = data[0]
+        self.max_val = data[1]
+        self.update_mask_array()
+
+    def get_mask_arrays(self):
+        return self._masked_arrays
+
+    def update_mask_array(self):
+        # TODO: rename apply_threshold_mask since its purpose has changed now?
+        # FIXME: Function parameters changed
+        self._masked_arrays = apply_threshold_mask(self.values)
+
+    def serialize(self):
+        return {
+            'min_val': self.min_val,
+            'max_val': self.max_val,
+            'name': self.name,
+            'mtype': self.mask_type,
+            'visible': self.mask_image,
+            'border': self.show_border,
+        }
+
+    def deserialize(self, data):
+        self.name = data['name']
+        self.mask_type = data['mtype']
+        self.mask_image = data['visible']
+        self.show_border = data['border']
+        self.set_data([data['min_val'], data['max_val']])
+
+
+class MaskManager(metaclass=Singleton):
+    def __init__(self, view_mode):
+        self.masks = {}
+        self.view_mode = view_mode
+
+    @property
+    def visible_masks(self):
+        return [k for k, v in self.masks if v.mask_image]
+
+    @property
+    def visible_boundaries(self):
+        return [k for k, v in self.masks if v.show_border]
+
+    @property
+    def threshold_mask(self):
+        for mask in self.masks.values():
+            if mask.mask_type == 'threshold':
+                return mask
+        return None
+
+    def add_mask(self, name, data, mtype, mask_image=True, show_border=False):
+        # Enforce name uniqueness
+        name = unique_name(self.masks.keys(), name)
+        if mtype == 'threshold':
+            new_mask = ThresholdMask(name, mtype, mask_image)
+        else:
+            new_mask = RegionMask(name, mtype, mask_image, show_border)
+        new_mask.set_data(self.view_mode, data)
+        self.masks[name] = new_mask
+
+    def remove_mask(self, name):
+        self.masks.pop(name)
+
+    def write_all_masks(self, h5py_group=None):
+        d = {}
+        for name, mask_info in self.masks:
+            d[name] = mask_info.serialize()
+        if h5py_group:
+            self.write_masks_to_group(d, h5py_group)
+        else:
+            self.export_masks_to_file(d)
+
+    def save_state(self, h5py_group):
+        if 'masks' not in h5py_group:
+            h5py_group.create_group('masks')
+
+        self.write_all_masks(h5py_group['masks'])
+
+    def load_masks(self, h5py_group):
+        # TODO: Handle case of detector name mismatch (loading wrong mask file)
+        items = h5py_group.items()
+        if '_visible' in h5py_group.values():
+            # This is a file using the old format
+            items = load_old_mask_file(h5py_group)
```

Comment:
I would say we should keep all of the logic as to whether we should convert an old format or not in the other compatibility file. Here, you can just do this instead:

```python
items = old_mask_compatibility(h5py_group)
```

That lets us keep all compatibility logic in that separate file (including checking the version). That `old_mask_compatibility()` function would be responsible for either converting an old format or just returning the items the way that they are if it is in the current format.

And then we should also start keeping a version in the h5py group instead of checking for `_visible`. So maybe have a `_version = 2` attribute. If the `_version` is not present, then assume it is version 1 (which means it would require converting the `_visible`). That will make future logic simpler if we have to change the mask format again.

Reply:
Sounds good, updated